x/slashing/keeper.go

```diff
@@ -120,3 +126,50 @@ func (k Keeper) handleValidatorSignature(ctx sdk.Context, pubkey crypto.PubKey,
 	// Set the updated signing info
 	k.setValidatorSigningInfo(ctx, address, signInfo)
 }
+
+// AddValidators adds the validators to the keepers validator addr to pubkey mapping.
+func (k Keeper) AddValidators(ctx sdk.Context, vals []abci.Validator) {
+	for i := 0; i < len(vals); i++ {
+		val := vals[i]
+		pubkey, err := tmtypes.PB2TM.PubKey(val.PubKey)
+		if err != nil {
+			continue
+		}
+		k.addPubkey(ctx, pubkey, false)
+	}
+}
+
+// TODO: Make a method to remove the pubkey from the map when a validator is unbonded.
+func (k Keeper) addPubkey(ctx sdk.Context, pubkey crypto.PubKey, del bool) {
+	addr := pubkey.Address()
+	if del {
+		k.deleteAddrPubkeyRelation(ctx, addr)
+	} else {
+		k.setAddrPubkeyRelation(ctx, addr, pubkey)
+	}
+}
+
+func (k Keeper) getPubkey(ctx sdk.Context, address crypto.Address) (crypto.PubKey, error) {
+	store := ctx.KVStore(k.storeKey)
+	var pubkey crypto.PubKey
+	err := k.cdc.UnmarshalBinary(store.Get(getAddrPubkeyRelationKey(address)), &pubkey)
```

Comment:
`MustUnmarshalBinary`

Reply:
(unless you wanted the custom error, in which case disregard)

Reply:
I do think a custom error message is preferrable